deeplearning4j-nn/src/main/java/org/deeplearning4j/datasets/iterator/AsyncDataSetIterator.java

```diff
@@ -385,7 +384,7 @@ public void remove() {
         private BlockingQueue<DataSet> queue;
         private DataSetIterator iterator;
         private DataSet terminator;
-        private AtomicBoolean isShutdown = new AtomicBoolean(false);
+        private boolean isShutdown = false; // locked around `this`
```

Comment:
This doesn't seem like a great idea to me?
We've gone from a lock free thread safe implementation, to something that's not lock free and not strictly thread safe? i.e., need volatile keyword here, at the very least?

Reply:
I am glad you brought this up.  There is a LOT of polling designs in this code base.  This was an example of where I switched a polling design to a `wait`/`notify`.

First this is correct, there is no need for a `volatile` keyword because this is only updated or read from inside a synchronized block of `this`.  If this variable needed to be read without synchronized, then we could add `volatile` to achieve that, but right now it's not necessary.

As for no-locking to locking, you are sorta correct.  The synchronized blocks will be held very short, only held while they are updating the value, and sending the notify call.  Or held while checking the value and waking up from their wait.  In this specific case, we will only have one thread doing the update.  And likely only one checking the update, but we can even assume more and this still should perform better than before.

When this would be problematic is if we have a lot of threads in the `wait`.  When the notify occurs those threads would need to acquire the lock one by one as they wake up, then release it.  So I would not suggest changing it to a `wait` / `notify` if there is going to be a lot of waiters, or high thread contention, I just don't think that is the case here.

What's wrong with polling you might ask?  Several things, but the biggest in my mind is the context switching that must occur.  The thread must wakeup, load it's state into the cpu cache, and then check a small boolean before it must again `park` itself.  The more threads you have doing this, the more thrashing occurs.

In short, no solution is the right solution for all cases.  I would like to provide future PR's with benchmarks and prove performance gains.  But these seemed like some obvious and safe points where I could start those changes.

Let me know if you have any more questions or thoughts about this.

Reply:
Thanks for the explanation, I think I'm happy with that. 👍 
Thinking about this more, given we really only have 2 threads operating here, we shouldn't really be worried about performance and lock contention anyway. And, as you say, context switching is a potential (likely?) performance issue anyway with the current design.